### DISCUSSION TITLE
CTR_EL0.IDC value under UTM is zero in HVF mode.

### DISCUSSION THREAD

**Discussion by pmsjt** [utmapp/UTM#4909] [Ideas]
The CTR_EL0 register contains varied information about the processor's cache configuration and cache maintenance requirements. In the Apple M1, the real value of CTR_EL0.IDC=1. This indicates that, when it is necessary to make the I-Cache and D-Cache coherent, it is not necessary to perform D-Cache invalidate operations to the Point-of-Unification.

I am observing that when running a VM under UTM, CTR_EL0.IDC=0. This has performance implications, as the underlying operating system and applications will loop around performing necessary cache maintenance operations. Frameworks like Java and .net, as well as x86/x64 emulation are particularly impacted by this as they generate executable code as they run (Just-In-Time).

Under Parallels I am observing the same behavior: CTR_EL0.IDC=0. However, under Apple's own VMM, CTR_EL0.IDC=1, which is the correct and optimal value.

This is not, strictly speaking, a QEMU thing either. I also experimented running a VM under QEMU + KVM on Linux (Ubuntu 22.04) using an Ampere Altra host. Just like on the M1, the Ampera Altra has CTR_EL0.IDC=1. In this setup VMs do read CTR_EL0.IDC=1, so QEMU with KVM is passing through the host value.

**Comment by osy** [comment]
CC @agraf
